### ISSUE TITLE
mTLS Authentication Broken on Node v15.x / Node v14.15.4

### ISSUE THREAD

**Issue by moloch--** [grpc/grpc-node#1669]
### Problem description

Using a custom CA to sign mTLS grpc-js connections results in an error on Node v15, this appears to be a change to how Node handles certificate validation in v15. Using the same code on v14.x results in a successful connection.

### Reproduction steps

Setup a mTLS authenticated grpc-js connection using Node v15.5.x

```js
rpcCredentials(): grpc.ChannelCredentials {
  const ca = Buffer.from(this._config.ca_certificate);
  const privateKey = Buffer.from(this._config.private_key);
  const certificate = Buffer.from(this._config.certificate);
  return grpc.credentials.createSsl(ca, privateKey, certificate, {
    checkServerIdentity: () => undefined,
  });
}
const rpc = new foo.RPCClient(this.rpcHost(), this.rpcCredentials(), {
  'grpc.max_send_message_length': 2 * Gb,
  'grpc.max_receive_message_length': 2 * Gb,
});
```

With `GRPC_TRACE all` and `GRPC_VERBOSITY DEBUG` enabled:

```
2021-01-22T16:59:13.964Z | resolving_load_balancer | dns:example.com:31337 IDLE -> IDLE
2021-01-22T16:59:13.965Z | connectivity_state | dns:example.com:31337 IDLE -> IDLE
2021-01-22T16:59:13.965Z | dns_resolver | Resolver constructed for target dns:example.com:31337
2021-01-22T16:59:13.967Z | channel | dns:example.com:31337 createCall [0] method="/foo.RPC/GetVersion", deadline=Infinity
2021-01-22T16:59:13.968Z | call_stream | [0] Sending metadata
2021-01-22T16:59:13.968Z | channel | Pick result: QUEUE subchannel: undefined status: undefined undefined
2021-01-22T16:59:13.968Z | call_stream | [0] write() called with message of length 0
2021-01-22T16:59:13.969Z | call_stream | [0] end() called
2021-01-22T16:59:13.969Z | call_stream | [0] deferring writing data chunk of length 5
2021-01-22T16:59:13.969Z | dns_resolver | Resolution update requested for target dns:example.com:31337
2021-01-22T16:59:13.970Z | resolving_load_balancer | dns:example.com:31337 IDLE -> CONNECTING
2021-01-22T16:59:13.971Z | channel | Pick result: QUEUE subchannel: undefined status: undefined undefined
2021-01-22T16:59:13.971Z | connectivity_state | dns:example.com:31337 IDLE -> CONNECTING
2021-01-22T16:59:13.971Z | resolving_load_balancer | dns:example.com:31337 CONNECTING -> CONNECTING
2021-01-22T16:59:13.971Z | channel | Pick result: QUEUE subchannel: undefined status: undefined undefined
2021-01-22T16:59:13.971Z | connectivity_state | dns:example.com:31337 CONNECTING -> CONNECTING
2021-01-22T16:59:14.020Z | dns_resolver | Resolved addresses for target dns:example.com:31337: [1.2.3.4:31337]
2021-01-22T16:59:14.021Z | pick_first | Connect to address list 1.2.3.4:31337
2021-01-22T16:59:14.022Z | subchannel_refcount | 1.2.3.4:31337 refcount 0 -> 1
2021-01-22T16:59:14.022Z | subchannel_refcount | 1.2.3.4:31337 refcount 1 -> 2
2021-01-22T16:59:14.022Z | pick_first | Start connecting to subchannel with address 1.2.3.4:31337
2021-01-22T16:59:14.022Z | pick_first | IDLE -> CONNECTING
2021-01-22T16:59:14.022Z | resolving_load_balancer | dns:example.com:31337 CONNECTING -> CONNECTING
2021-01-22T16:59:14.023Z | channel | Pick result: QUEUE subchannel: undefined status: undefined undefined
2021-01-22T16:59:14.023Z | connectivity_state | dns:example.com:31337 CONNECTING -> CONNECTING
2021-01-22T16:59:14.023Z | subchannel | 1.2.3.4:31337 IDLE -> CONNECTING
2021-01-22T16:59:14.024Z | pick_first | CONNECTING -> CONNECTING
2021-01-22T16:59:14.024Z | resolving_load_balancer | dns:example.com:31337 CONNECTING -> CONNECTING
2021-01-22T16:59:14.024Z | channel | Pick result: QUEUE subchannel: undefined status: undefined undefined
2021-01-22T16:59:14.024Z | connectivity_state | dns:example.com:31337 CONNECTING -> CONNECTING
2021-01-22T16:59:14.160Z | subchannel | 1.2.3.4:31337 connection closed with error self signed certificate
2021-01-22T16:59:14.160Z | subchannel | 1.2.3.4:31337 CONNECTING -> TRANSIENT_FAILURE
2021-01-22T16:59:14.160Z | pick_first | CONNECTING -> TRANSIENT_FAILURE
2021-01-22T16:59:14.160Z | resolving_load_balancer | dns:example.com:31337 CONNECTING -> TRANSIENT_FAILURE
2021-01-22T16:59:14.160Z | channel | Pick result: TRANSIENT_FAILURE subchannel: undefined status: 14 No connection established
2021-01-22T16:59:14.160Z | call_stream | [0] cancelWithStatus code: 14 details: "No connection established"
2021-01-22T16:59:14.161Z | call_stream | [0] ended with status: code=14 details="No connection established"
2021-01-22T16:59:14.161Z | connectivity_state | dns:example.com:31337 CONNECTING -> TRANSIENT_FAILURE

Error: 14 UNAVAILABLE: No connection established
    at Object.callErrorFromStatus (/Users/user/git/foobar/node_modules/@grpc/grpc-js/build/src/call.js:31:26)
    at Object.onReceiveStatus (/Users/user/git/foobar/node_modules/@grpc/grpc-js/build/src/client.js:176:52)
    at Object.onReceiveStatus (/Users/user/git/foobar/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:334:141)
    at Object.onReceiveStatus (/Users/user/git/foobar/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:297:181)
    at /Users/user/git/foobar/node_modules/@grpc/grpc-js/build/src/call-stream.js:130:78
    at processTicksAndRejections (node:internal/process/task_queues:76:11) {
  code: 14,
  details: 'No connection established',
  metadata: Metadata { internalRepr: Map(0) {}, options: {} }
}
```

### Environment
 - MacOS Big Sur
 - Node version v15.5.x
 - Node installation method: nvm
 - grpc/grpc-js@1.2.4

**Comment by murgatroid99** [comment]
If a change in Node broke code that was previously working, that seems like a Node issue to me. Have you filed an issue against Node itself?

**Comment by moloch--** [comment]
This seems like something I'd expect grpc-js to abstract since I'm providing the TLS configuration to grpc-js, but maybe I'm wrong? From what I can see this was a deliberate change in Node v15 to make it harder to use self-signed CAs.

**Comment by moloch--** [comment]
For context it seems like Node v15 is rejecting the certificate before validating it against the CA I provided to gPRC as credential.

**Comment by moloch--** [comment]
After additional investigation the code seems to work in v14.15.3 but not v14.15.4

The [v14.15.4](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V14.md#14.15.4) notes some changes to the TLS code but it's not clear to me why it would break gRPC custom CAs.

**Comment by moloch--** [comment]
Opened https://github.com/nodejs/node/issues/37025

**Comment by murgatroid99** [comment]
grpc-js just delegates all TLS operations to Node's own TLS module. There isn't really any abstraction. If the result of the Node issue you filed is that there is an option to set to get the same behavior as in earlier versions, then I can look into setting that automatically, but if the new version of OpenSSL just has different certificate validation behavior, there isn't much I can do about that.

**Comment by moloch--** [comment]
@murgatroid99 is there any way to get more detailed logs than using `GRPC_VERBOSITY` and `GRPC_TRACE`? The only error i can pull out is `subchannel | 127.0.0.1:31337 connection closed with error self signed certificate` but it's tricky to find where this is coming from in the `grpc-js` code / which node API is returning the error.

**Comment by moloch--** [comment]
@murgatroid99 I've created a standalone repo using the gRPC "Hello World" and only adding mutual-TLS authenticated.

https://github.com/moloch--/mtls-broken

It does not appear to be currently possible to connect from a grpc-js client to a Go-based gRPC server using mutual-TLS.

**Comment by murgatroid99** [comment]
I can tell you that that trace line comes from [this line](https://github.com/grpc/grpc-node/blob/master/packages/grpc-js/src/subchannel.ts#L448). That is handling an `error` event emitted by an [`Http2Session` object](https://nodejs.org/api/http2.html#http2_event_error).

You can get additional http2 debugging information with the environment variable `NODE_DEBUG=http2`, and additional TLS debugging information with the Node command line argument [`--trace-tls`](https://nodejs.org/api/cli.html#cli_trace_tls).

**Comment by moloch--** [comment]
Thanks for the tips, looks like it's related to:

```
HTTP2 82779: Http2Session client: start closing/destroying Error: self signed certificate
    at TLSSocket.onConnectSecure (node:_tls_wrap:1531:34)
    at TLSSocket.emit (node:events:378:20)
    at TLSSocket._finishInit (node:_tls_wrap:945:8)
    at TLSWrap.ssl.onhandshakedone (node:_tls_wrap:719:12) {
  code: 'DEPTH_ZERO_SELF_SIGNED_CERT'
}
```

**Comment by moloch--** [comment]
I get the same error when I replace the client code's CA with `asdf` so Node isn't even attempting to validate the certificate using the provided authority: https://github.com/moloch--/mtls-broken/blob/ca/client/helloworld-client.js#L11

**Comment by moloch--** [comment]
@murgatroid99  and any that finds this thread, after must more debugging I identified the problem.

In Node versions up to and including v14.15.3 Node will allow a CA to have a blank CN, however Node versions v14.15.4 and later, Node will reject any CA with a blank CN with no other considerations given to the certificate. To fix for v14.15.4 and later simply provide any non-blank CN value, it's completely arbitrary it'll take an non-empty value e.g. `a`. Yup.
